core/src/main/java/org/elasticsearch/rest/BaseRestHandler.java

```diff
@@ -43,4 +50,52 @@ protected BaseRestHandler(Settings settings) {
         super(settings);
         this.parseFieldMatcher = new ParseFieldMatcher(settings);
     }
+
+    @Override
+    public final void handleRequest(RestRequest request, RestChannel channel, NodeClient client) throws Exception {
+        // prepare the request for execution; has the side effect of touching the request parameters
+        final Runnable action = doRequest(request, channel, client);
+
+        // validate unconsumed params, but we must exclude params used to format the response
+        // we copy because we do not want to modify the request params
+        final Set<String> unconsumedParams = new HashSet<>(request.unconsumedParams());
+        final Set<String> responseParams = responseParams();
+        final Iterator<String> it = unconsumedParams.iterator();
+
+        // this has to use an iterator lest a ConcurrentModificationException will arise
+        while (it.hasNext()) {
+            final String unconsumedParam = it.next();
+            if (responseParams.contains(unconsumedParam)) {
+                it.remove();
+            }
+        }
+
+        // validate the non-response params
+        if (!unconsumedParams.isEmpty()) {
+            throw new IllegalArgumentException("request [" + request.path() + "] contains unused params: " + unconsumedParams.toString());
+        }
+
+        // execute the action
+        action.run();
+    }
+
+    /**
+     * Prepare the request for execution. Implementations should consume all request params before
+     * returning the runnable for actual execution. Unconsumed params will immediately terminate
+     * execution of the request. However, some params are only used in processing the response;
+     * implementations can override {@link BaseRestHandler#responseParams()} to indicate such
+     * params.
+     *
+     * @param request the request to execute
+     * @param channel the channel for sending the response
+     * @param client  client for executing actions on the local node
+     * @return the action to execute
+     * @throws Exception if an exception occurred preparing the action for exectuion
+     */
+    protected abstract Runnable doRequest(RestRequest request, RestChannel channel, NodeClient client) throws Exception;
+
+    protected Set<String> responseParams() {
```

Comment:
Sure, I can do that. There was a comment on the Javadoc for `BaseRestHandler#doRequest` in this regard, but I agree that it would be nice to have it here too.